### DISCUSSION TITLE
Strange compilation issue, thread Sleep doesn't work

### DISCUSSION THREAD

**Discussion by 28106830** [nanoframework/Home#1576] [General]
I downloaded today's latest nf interpreter source code (the same goes for the previous issue),
I compiled using VSCode under Windows and selected ESP32-S3. However, without moving anything, the generated firmware had the following issues:
When Sleeping (10 * 1000), the thread should pause for 10 seconds, but in reality, it comes out directly
This situation is not always present, but the probability is high
3. Especially after connecting to WiFi and establishing a network connection
When sleeping normally, the Sleep process will exit directly after sending data through the serial port, as if awakened
So far, all the firmware I have compiled has this issue
6. May I ask if you can provide the sdkconfig file for the firmware generated on your end? The microcontroller I am currently using is ESP32-S3

**Comment by alberk8** [comment]
When you change anything in the [sdkconfig](https://github.com/nanoframework/nf-interpreter/tree/main/targets/ESP32/_IDF) then you need to delete the build directory and the sdkconfig file in the root directory of nf-interpreter before building again otherwise funny things will happen.

**Reply by josesimoes** [reply]
We don't have any private sdkconfig file. That is generated on each build, using the config options on the repo.

**Reply by 28106830** [reply]
Thank you very much for your reply
My operating steps are:
1. Download the latest source code
2. Open with VSCode on Windows 11
3. Open ESP32 using a container
4. Copy two files from the config and change the 'user tools repos container' in the 'user tools repos. json' file to 'user tools repos container'
5. Restart
6. Change project configuration to ESP32_S3
7. Restart
8. Project Generation
9. Close the project, delete Build and sdkconfig
10. Regenerate
But there is still a significant difference between the generated file and the automatically downloaded one. Using a comparison file tool, basically all the previous ones are different
I have attached my sdkconfig file. If you have time, you can also attach the sdkconfig file so that I can make a comparison
![image](https://githu
[sdkconfig.zip](https://github.com/user-attachments/files/18273489/sdkconfig.zip)
b.com/user-attachments/assets/7d8bc2f5-a1f5-4f37-9b05-51af8a8ed01d)
I did not make any modifications to the source code

**Comment by 28106830** [comment]
I just copied user-prefs.json without making any changes
I just copied the template and changed the "name" from "user_tools_repos_container" to "name" from "user_tools_repos",
Does this file also need to be modified?

**Reply by josesimoes** [reply]
Make sure to strictly follow the instructions [here](https://docs.nanoframework.net/content/building/using-dev-container.html).
And adjust the user preferences in the json config files. 
I don't understand exactly what you're trying to compare between the local build and the one built on the pipeline... but know that build in the pipeline is set for `MinRelSize`.

**Comment by 28106830** [comment]
I see that the version numbers are all different. This is in user-prefs.json. What other changes have been made?

**Reply by josesimoes** [reply]
Version number on the pipeline is computed by nbgv tool with commit history. Local build has the build number coming from the config file.

**Comment by 28106830** [comment]
Set : "CMAKE_BUILD_TYPE": "Release",
build]   216 |     hr = g_CLR_RT_ExecutionEngine.NewObjectFromIndex(*returnArray, rmtCommandTypeDef);
[build]       |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] /workspaces/nf-interpreter-main/targets/ESP32/_nanoCLR/nanoFramework.Hardware.Esp32.Rmt/nanoFramework_hardware_esp32_rmt_native_nanoFramework_Hardware_Esp32_Rmt_ReceiverChannel.cpp: In static member function 'static HRESULT Library_nanoFramework_hardware_esp32_rmt_native_nanoFramework_Hardware_Esp32_Rmt_ReceiverChannel::NativeRxGetAllItems___SZARRAY_nanoFrameworkHardwareEsp32RmtRmtCommand(CLR_RT_StackFrame&)':
[build] /workspaces/nf-interpreter-main/targets/ESP32/_nanoCLR/nanoFramework.Hardware.Esp32.Rmt/nanoFramework_hardware_esp32_rmt_native_nanoFramework_Hardware_Esp32_Rmt_ReceiverChannel.cpp:312:31: note: 'arrayDataPtr' was declared here
[build]   312 |             CLR_RT_HeapBlock *arrayDataPtr;
[build]       |                               ^~~~~~~~~~~~

**Comment by 28106830** [comment]
I need to add some firmware features, but I haven't reached this stage yet because I am currently unable to compile firmware that is consistent with the official version
Of course, the firmware will include some dates, so it's okay if they are inconsistent
I found that the firmware I generated in nanoCLR.bin contains the ESP-IDF version, but the official version number is not available!
I generated it using a container and strictly followed the operation steps. You mentioned that you need to modify user-prefs.json. Would it be convenient to provide a copy as a reference
It would be even better if there were user tools-repos.json and sdkconfig
The firmware issue I generated:
1. The file is relatively large
2. Without simulation, there is a high probability that WiFi cannot be connected
3. All programs use Thread The position of Sleep is invalid, and the result is that the log keeps printing, MQTT data is uploaded 3 times per second, and all the places where the collector needs to wait are directly sent over

**Comment by 28106830** [comment]
I found the problem
My previous sdkconfig was not deleted
There is a configuration inside that is set to VNet FRERTOS-HZ
If it is 100, then there is no problem
But if it is changed to 1000, then Thread Sleep will make mistakes

**Reply by josesimoes** [reply]
That makes sense. Sleep misbehaving on bad timing, has to do with clock. 
When changing configuration or target, sdkconfig has to be deleted and recreated.
